File: py/src/axon/hooks/hosts/opencode.py
"""The OpenCode codec.

OpenCode hooks are in-process JS/TS plugin callbacks, not subprocesses:
there is no exit-code contract (spec/hosts/opencode/host.yaml has no
exit_codes key). encode_decision always returns exit 0; decode_decision
ignores its exit argument and decodes the decision from stdout JSON only
(invariant 8).

Event names are never hardcoded here: they come from
capabilities().host_event for encoding and the native-wins decode map for
decoding, so spec/hosts/opencode/capabilities.yaml is the single source
of the canonical<->dotted event mapping.

Mirrors hooks/hosts/opencode/codec.go; matched against
ts/src/hooks/hosts/opencode.ts for shape.
"""

from __future__ import annotations

from axon.capabilities import Capabilities, host_event as cap_host_event, level, load_capabilities
from axon.errors import SchemaError, UnsupportedActionError, UnsupportedEventError
from axon.hooks.decision import Decision
from axon.hooks.input import Input

_KNOWN_INPUT_KEYS = {"type", "session_id", "tool", "input", "output", "prompt"}


class OpencodeCodec:
    def __init__(self) -> None:
        self._caps = load_capabilities("opencode")
        self._rev = _decode_map(self._caps)

    def host(self) -> str:
        return "opencode"

    def capabilities(self) -> Capabilities:
        return self._caps

    def encode_input(self, inp: Input) -> dict:
        native = cap_host_event(self._caps, inp.event)
        if native is None:
            raise UnsupportedEventError("opencode", inp.event)

        out: dict = dict(inp.extra or {})
        out["type"] = native
        out["session_id"] = inp.session_id
        if inp.event in ("PreToolUse", "PermissionRequest"):
            out["tool"] = inp.tool_name
            out["input"] = inp.tool_input
        elif inp.event == "PostToolUse":
            out["tool"] = inp.tool_name
            out["input"] = inp.tool_input
            out["output"] = inp.tool_response
        elif inp.event == "UserPromptSubmit":
            out["prompt"] = inp.prompt
        return out

    def decode_input(self, raw: dict) -> Input:
        m = _as_dict(raw, "opencode input")
        native = m.get("type")
        if not isinstance(native, str) or native == "":
            # A missing discriminator is a malformed payload, not an
            # unsupported event: reporting it as "opencode/" named no
            # event at all.
            raise SchemaError("opencode input: missing type")
        event = self._rev.get(native)
        if event is None:
            raise UnsupportedEventError("opencode", native)

        extra = {k: v for k, v in m.items() if k not in _KNOWN_INPUT_KEYS}

        inp = Input(
            event=event,
            session_id=m.get("session_id") if isinstance(m.get("session_id"), str) else "",
            cwd="",
            extra=extra,
        )
        if isinstance(m.get("tool"), str):
            inp.tool_name = m["tool"]
        if isinstance(m.get("input"), dict):
            inp.tool_input = m["input"]
        if isinstance(m.get("output"), dict):
            inp.tool_response = m["output"]
        if isinstance(m.get("prompt"), str):
            inp.prompt = m["prompt"]
        return inp

    # encode_decision always returns exit 0 (invariant 8). Warn is passed
    # through verbatim as {"action":"warn","message":...} -- folding it
    # to block would change semantics (block a tool the handler only
    # warned about), so it is never done (invariant 2).
    def encode_decision(self, d: Decision) -> tuple[object, int]:
        self._check_event(d.event)
        if d.action == "allow":
            return {"action": "allow"}, 0
        if d.action == "rewrite":
            # Field name "rewrite" per xat
            # spec/hook-output/opencode/tool.execute.before-rewrite.json.
            return {"action": "rewrite", "rewrite": d.rewrite}, 0
        if d.action == "warn":
            return {"action": "warn", "message": d.message or ""}, 0
        if d.action == "block":
            return {"action": "block", "message": d.message or ""}, 0
        raise UnsupportedActionError(f"opencode cannot express action {d.action!r}")

    # decode_decision ignores exit entirely: OpenCode decisions live in
    # stdout JSON only. Empty stdout decodes to allow.
    def decode_decision(self, stdout: object, exit_code: int) -> Decision:
        if _is_empty_stdout(stdout):
            return Decision(action="allow")
        m = _as_dict(stdout, "opencode decision")
        d = Decision(action="allow")
        action = m.get("action")
        if action == "block":
            d.action = "block"
        elif action == "warn":
            d.action = "warn"
        elif action == "rewrite":
            d.action = "rewrite"
            d.rewrite = m.get("rewrite")
        message = m.get("message")
        d.message = message if isinstance(message, str) else ""
        return d

    # _check_event rejects a decision naming an event OpenCode has no
    # wire shape for. A blank event keeps the host's default shape.
    def _check_event(self, ev: str) -> None:
        if not ev:
            return
        lvl, ok = level(self._caps, ev)
        if not ok or lvl not in ("native", "close"):
            raise UnsupportedEventError("opencode", ev)


# _decode_map: native rows always win over close rows. A close row whose
# host event no native row claims is unambiguous and does decode
# (todo.updated -> TaskCompleted and tui.prompt.append ->
# UserPromptSubmit are the only way those envelopes decode at all).
def _decode_map(caps: Capabilities) -> dict[str, str]:
    m: dict[str, str] = {}
    for p in caps.native:
        if p.host_event in m:
            raise SchemaError(
                f"{caps.host}: host event {p.host_event} maps to both {m[p.host_event]} and {p.event} under native"
            )
        m[p.host_event] = p.event
    seen_close: dict[str, str] = {}
    for p in caps.close or []:
        if p.host_event in seen_close:
            raise SchemaError(
                f"{caps.host}: host event {p.host_event} maps to both "
                f"{seen_close[p.host_event]} and {p.event} under close"
            )
        seen_close[p.host_event] = p.event
        if p.host_event in m:
            continue
        m[p.host_event] = p.event
    return m


def _as_dict(v: object, what: str) -> dict:
    if not isinstance(v, dict):
        raise SchemaError(f"{what}: expected a JSON object")
    return v


def _is_empty_stdout(stdout: object) -> bool:
    if stdout is None or stdout == "":
        return True
    if isinstance(stdout, str):
        return stdout.strip() == ""
    return False


def make_codec() -> OpencodeCodec:
    return OpencodeCodec()

File: ts/src/hooks/hosts/opencode.ts
import { loadCapabilities, level, hostEvent as capHostEvent, type Capabilities } from "../../capabilities";
import type { Event } from "../../events";
import { UnsupportedActionError, UnsupportedEventError, SchemaError } from "../../errors";
import type { Codec } from "../codec";
import type { Input } from "../input";
import type { Decision } from "../decision";

/**
 * The OpenCode codec.
 *
 * OpenCode hooks are in-process JS/TS plugin callbacks, not subprocesses:
 * there is no exit-code contract (spec/hosts/opencode/host.yaml has no
 * exit_codes key). encodeDecision always returns exit 0; decodeDecision
 * ignores its exit argument and decodes the decision from stdout JSON
 * only (invariant 8).
 *
 * Event names are never hardcoded here: they come from
 * capabilities().hostEvent for encoding and the native-wins decode map
 * for decoding, so spec/hosts/opencode/capabilities.yaml is the single
 * source of the canonical<->dotted event mapping.
 */
class OpencodeCodec implements Codec {
  private readonly caps: Capabilities;
  private readonly rev: Map<string, Event>;

  constructor() {
    this.caps = loadCapabilities("opencode");
    this.rev = decodeMap(this.caps);
  }

  host(): string {
    return "opencode";
  }

  capabilities(): Capabilities {
    return this.caps;
  }

  encodeInput(input: Input): unknown {
    const native = capHostEvent(this.caps, input.event);
    if (native === undefined) {
      throw new UnsupportedEventError("opencode", input.event);
    }

    const out: Record<string, unknown> = { ...(input.extra ?? {}) };
    out.type = native;
    out.session_id = input.sessionId;
    switch (input.event) {
      case "PreToolUse":
      case "PermissionRequest":
        out.tool = input.toolName;
        out.input = input.toolInput;
        break;
      case "PostToolUse":
        out.tool = input.toolName;
        out.input = input.toolInput;
        out.output = input.toolResponse;
        break;
      case "UserPromptSubmit":
        out.prompt = input.prompt;
        break;
      default:
        break;
    }
    return out;
  }

  decodeInput(raw: unknown): Input {
    const m = asRecord(raw, "opencode input");
    const native = m.type;
    if (typeof native !== "string" || native === "") {
      throw new SchemaError("opencode input: missing type");
    }
    const event = this.rev.get(native);
    if (event === undefined) {
      throw new UnsupportedEventError("opencode", native);
    }

    const known = new Set(["type", "session_id", "tool", "input", "output", "prompt"]);
    const extra: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(m)) {
      if (!known.has(k)) extra[k] = v;
    }

    const input: Input = {
      event,
      sessionId: typeof m.session_id === "string" ? m.session_id : "",
      cwd: "",
      extra,
    };
    if (typeof m.tool === "string") input.toolName = m.tool;
    if (isRecord(m.input)) input.toolInput = m.input;
    if (isRecord(m.output)) input.toolResponse = m.output;
    if (typeof m.prompt === "string") input.prompt = m.prompt;
    return input;
  }

  // encodeDecision always returns exit 0 (invariant 8). Warn is passed
  // through verbatim as {"action":"warn","message":...} -- folding it to
  // block would change semantics (block a tool the handler only warned
  // about), so it is never done (invariant 2).
  encodeDecision(d: Decision): { stdout: unknown; exit: number } {
    this.checkEvent(d.event);
    switch (d.action) {
      case "allow":
        return { stdout: { action: "allow" }, exit: 0 };
      case "rewrite":
        return { stdout: { action: "rewrite", rewrite: d.rewrite }, exit: 0 };
      case "warn":
        return { stdout: { action: "warn", message: d.message ?? "" }, exit: 0 };
      case "block":
        return { stdout: { action: "block", message: d.message ?? "" }, exit: 0 };
      default:
        throw new UnsupportedActionError(`opencode cannot express action ${String(d.action)}`);
    }
  }

  // decodeDecision ignores exit entirely: OpenCode decisions live in
  // stdout JSON only. Empty stdout decodes to allow.
  decodeDecision(stdout: unknown, _exit: number): Decision {
    if (isEmptyStdout(stdout)) {
      return { action: "allow" };
    }
    const m = asRecord(stdout, "opencode decision");
    const d: Decision = { action: "allow" };
    switch (m.action) {
      case "block":
        d.action = "block";
        break;
      case "warn":
        d.action = "warn";
        break;
      case "rewrite":
        d.action = "rewrite";
        d.rewrite = m.rewrite;
        break;
      default:
        break;
    }
    d.message = typeof m.message === "string" ? m.message : "";
    return d;
  }

  // checkEvent rejects a decision naming an event OpenCode has no wire
  // shape for. A blank event keeps the host's default shape.
  private checkEvent(ev: Event | undefined): void {
    if (ev === undefined || ev === "") return;
    const { level: lvl, ok } = level(this.caps, ev);
    if (!ok || (lvl !== "native" && lvl !== "close")) {
      throw new UnsupportedEventError("opencode", ev);
    }
  }
}

// decodeMap: native rows always win over close rows. A close row whose
// host event no native row claims is unambiguous and does decode
// (todo.updated -> TaskCompleted and tui.prompt.append -> UserPromptSubmit
// are the only way those envelopes decode at all).
function decodeMap(caps: Capabilities): Map<string, Event> {
  const m = new Map<string, Event>();
  for (const p of caps.native) {
    if (m.has(p.hostEvent)) {
      throw new SchemaError(`${caps.host}: host event ${p.hostEvent} maps to both ${m.get(p.hostEvent)} and ${p.event} under native`);
    }
    m.set(p.hostEvent, p.event);
  }
  const seenClose = new Map<string, Event>();
  for (const p of caps.close ?? []) {
    if (seenClose.has(p.hostEvent)) {
      throw new SchemaError(`${caps.host}: host event ${p.hostEvent} maps to both ${seenClose.get(p.hostEvent)} and ${p.event} under close`);
    }
    seenClose.set(p.hostEvent, p.event);
    if (m.has(p.hostEvent)) continue;
    m.set(p.hostEvent, p.event);
  }
  return m;
}

function isRecord(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

function asRecord(v: unknown, what: string): Record<string, unknown> {
  if (!isRecord(v)) throw new SchemaError(`${what}: expected a JSON object`);
  return v;
}

function isEmptyStdout(stdout: unknown): boolean {
  if (stdout === undefined || stdout === null || stdout === "") return true;
  if (typeof stdout === "string") return stdout.trim() === "";
  return false;
}

export function makeCodec(): Codec {
  return new OpencodeCodec();
}
